### PR TITLE
8329529: Serial: Move _saved_mark_word out of ContiguousSpace

### PR DESCRIPTION
**Pull request by lgxbslgx**
Hi all,

This patch removes the field `ContiguousSpace::_saved_mark_word` and its related methods, adds the corresponding fields to `DefNewGeneration` and `TenuredGeneration`, and moves the method `ContiguousSpace::oop_since_save_marks_iterate` to `Generation`. Many usages are adjusted as well.

An alternative way is add a class which extends the `ContiguousSpace` to place `_saved_mark_word` and its related methods. But this way would increase the class hierarchy. Not really know whether it deserves to do that.

The tests `make test-tier1_gc` passed locally. Thanks for taking the time to review.

Best Regards,
-- Guoxiong

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8329529](https://bugs.openjdk.org/browse/JDK-8329529): Serial: Move _saved_mark_word out of ContiguousSpace (**Enhancement** - P4)

### Reviewers
 * [Albert Mingkun Yang](https://openjdk.org/census#ayang) (@albertnetymk - **Reviewer**) ⚠️ Review applies to [8ef588b5](https://git.openjdk.org/jdk/pull/18603/files/8ef588b5c110d1a037369da8875bb964f51ca631)
 * [Ivan Walulya](https://openjdk.org/census#iwalulya) (@walulyai - **Reviewer**) ⚠️ Review applies to [8ef588b5](https://git.openjdk.org/jdk/pull/18603/files/8ef588b5c110d1a037369da8875bb964f51ca631)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/18603/head:pull/18603` \
`$ git checkout pull/18603`

Update a local copy of the PR: \
`$ git checkout pull/18603` \
`$ git pull https://git.openjdk.org/jdk.git pull/18603/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 18603`

View PR using the GUI difftool: \
`$ git pr show -t 18603`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/18603.diff">https://git.openjdk.org/jdk/pull/18603.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/18603#issuecomment-2034439271)

### CODE REVIEW

File: src/hotspot/share/gc/serial/defNewGeneration.hpp
Comment (albertnetymk) [inline src/hotspot/share/gc/serial/defNewGeneration.hpp:143]: Are they really needed? I'd expect `saved_mark_word` to be useful only for copy-destination, i.e. to-space and old-gen.

File: src/hotspot/share/gc/serial/generation.hpp
Comment (lgxbslgx) [reply src/hotspot/share/gc/serial/generation.hpp]: Fixed.

File: src/hotspot/share/gc/serial/defNewGeneration.hpp
Comment (lgxbslgx) [reply src/hotspot/share/gc/serial/defNewGeneration.hpp]: Before, I tried to keep the previous meaning and didn't delete the existing code.

The `save_mark_word` of eden-space and from-space only influence the assertment. I removed them just now.

**Review by albertnetymk — APPROVED** [review]

**Comment by lgxbslgx** [comment]
@albertnetymk Thanks for your review. Waiting for another review.

File: src/hotspot/share/gc/serial/tenuredGeneration.hpp
Comment (walulyai) [inline src/hotspot/share/gc/serial/tenuredGeneration.hpp:69]: `// Saved mark word` seems unnecessary

File: src/hotspot/share/gc/serial/defNewGeneration.hpp
Comment (walulyai) [inline src/hotspot/share/gc/serial/defNewGeneration.hpp:174]: comment doesn't seem to add any value

**Review by walulyai — APPROVED** [review]
LGTM!

Minor:

File: src/hotspot/share/gc/serial/tenuredGeneration.hpp
Comment (lgxbslgx) [reply src/hotspot/share/gc/serial/tenuredGeneration.hpp]: Removed.

File: src/hotspot/share/gc/serial/defNewGeneration.hpp
Comment (lgxbslgx) [reply src/hotspot/share/gc/serial/defNewGeneration.hpp]: Removed.

**Comment by lgxbslgx** [comment]
@albertnetymk @walulyai Thanks for your reviews.

/integrate

**Pull request closed without merging**
